### ISSUE TITLE
Worker cannot be joined to swarm cluster

### ISSUE THREAD

**Issue by aleskurlovich** [moby/moby#34996]
**Description**
Worker cannot be joined to swarm manager.

**Steps to reproduce the issue:**
1. Create manager node:
```
docker swarm init
```
2. Create worker node:
```
docker swarm join --token SWMTKN-1-1m7h3jnhycq46r5qzunx8gwvjjr4f30ckfqet2uqgdj60x3a9a-7kr5m1c3syidtmpgouh54f0fo <ip>:2377
```

**Describe the results you received:**
```
Sep 27 13:07:18 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:18.304422384+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:20 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:20.307296581+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:22 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:22.309839445+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:24 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:24.312388683+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:26 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:26.315234191+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:28 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:28.315699672+02:00" level=error msg="cluster exited with error: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
Sep 27 13:07:28 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:28.316059555+02:00" level=error msg="Handler for POST /v1.32/swarm/join returned error: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
```

**Describe the results you expected:**
Worker is joined.

**Additional information you deem important (e.g. issue happens only occasionally):**
We faced to this issue after Docker was upgraded from 17.06.2-ce to 17.09.0-ce. Now it is not possible to join any worker to the swarm.

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:49 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 15
Server Version: 17.09.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: keosiajh7onqpn42u3ixsrmb3
 Is Manager: true
 ClusterID: qe8cswquhtjl817eiu4gtzyio
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.33.177.47
 Manager Addresses:
  <ip>:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.21.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 9.601GiB
Name: q4de3csye47
ID: 2FXY:XP3H:UWFL:3R7U:CF5H:THXJ:XC5K:SZTS:U4LV:AY7D:3JNV:FOJF
Docker Root Dir: /docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://<ip>:3128/
Https Proxy: http://<ip>:3128/
No Proxy: <ip>,registry
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 10.33.177.26:5000
 registry:5000
 127.0.0.0/8
Live Restore Enabled: false

WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
OS: Centos 7
Linux q4de3csye43 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**Comment by thaJeztah** [comment]
Given that you're using a proxy, this looks similar to https://github.com/moby/moby/issues/34825#issuecomment-331100937

Can you try if the solutions mentioned there help you resolve this?

**Comment by thaJeztah** [comment]
Marking this as bug for now, until it's properly triaged where this change came from

ping @nishanttotla @dperny PTAL

**Comment by Remigius2011** [comment]
just asking: is there any possibility to trace network traffic during swarm join (other than tcpdump - which is a bit brute force)? this might be handy in such situations.

**Comment by aleskurlovich** [comment]
Thank you. It was problem with proxy.

**Comment by markelhas** [comment]
@aleskurlovich how did u set the no proxy?
I've tested on /etc/environment and /etc/systemd/system/docker.service.d/http-proxy.conf, but for me keeps not being able to join the swarm.

**Comment by Prit1104** [comment]
@aleskurlovich i am having the problem while joining the node as a manager. can you explain what was the proxy settings that you resolved?

**Comment by thaJeztah** [comment]
Docker 18.09.4 has a fix for situations where a proxy is configured, and where it would attempt to use that proxy for connecting workers/managers to the swarm cluster; in 18.09.4 it should no longer be needed to explicitly exclude manager nodes using `NO_PROXY`.

**Comment by Huynhvuong** [comment]
I am using Docker for Mac stable 18.09.2. I can join swarm by some nodes which created by docker-machine. But I can't join swam by host Mac. Could you explain me @thaJetah.
Thank you very much

**Comment by thaJeztah** [comment]
Hard to tell just from that information; note that Docker for Mac was primarily designed to run standalone (and in a single-node swarm cluster); joining Docker for Mac to a cluster (or nodes _to_ a Docker for Mac cluster) may be complicated, depending on your setup

**Comment by Gopinathpalanisami** [comment]
Hi Sebastiaan van Stijn,

I need your help here as you said docker swarm join issue is fixed in 18.06 version only, Still am facing same issue with 18.09.6 version ; 

docker version : 18.09.6
But still facing same below issue. 
https://github.com/moby/moby/issues/34996 …

/etc/systemd/system/docker.service.d/http-proxy.conf
Environment="NO_PROXY=10.199.197.170,10.118.246.206"

Am not sure how to update NO_PROXY parameter .

Local Host IP :10.199.197.170
Remote Host IP :10.118.246.206

When I try to verify No output.
sles15:/etc/systemd/system/docker.service.d # systemctl show docker --property Environment
Environment=

sles15:~ # docker swarm join --token SWMTKN-1-0o7yvygjvcrh6rib47vqj1jjqe84ua4eo7mbyoloewx84brixa-7q7og2n197yjve9v2zj8lbwwx 10.118.246.206:2377
Error response from daemon: Timeout was reached before node joined. The attempt to join the swarm will continue in the background. Use the "docker info" command to see the current swarm status of your node.

In /var/log/messages

2019-07-10T10:01:26.858048-04:00 sles15 dockerd[25407]: time="2019-07-10T10:01:26.857759256-04:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" module=node
2019-07-10T10:01:33.858575-04:00 sles15 dockerd[25407]: time="2019-07-10T10:01:33.858412400-04:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" module=node
2019-07-10T10:01:40.859291-04:00 sles15 dockerd[25407]: time="2019-07-10T10:01:40.859070236-04:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" module=node

Please help me in docker swarm join issue.

**Comment by Gopinathpalanisami** [comment]
@thaJeztah ,

I need your help here as you said docker swarm join issue is fixed in 18.06 version only, Still am facing same issue; 

docker version : 18.09.6
But still facing same below issue. 
https://github.com/moby/moby/issues/34996 …

/etc/systemd/system/docker.service.d/http-proxy.conf
Environment="NO_PROXY=10.199.197.170,10.118.246.206"

Local Host IP :10.199.197.170
Remote Host IP :10.118.246.206

When I try to verify No output.
sles15:/etc/systemd/system/docker.service.d # systemctl show docker --property Environment
Environment=

sles15:~ # docker swarm join --token SWMTKN-1-0o7yvygjvcrh6rib47vqj1jjqe84ua4eo7mbyoloewx84brixa-7q7og2n197yjve9v2zj8lbwwx 10.118.246.206:2377
Error response from daemon: Timeout was reached before node joined. The attempt to join the swarm will continue in the background. Use the "docker info" command to see the current swarm status of your node.

In /var/log/messages

2019-07-10T10:01:26.858048-04:00 sles15 dockerd[25407]: time="2019-07-10T10:01:26.857759256-04:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" module=node
2019-07-10T10:01:33.858575-04:00 sles15 dockerd[25407]: time="2019-07-10T10:01:33.858412400-04:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" module=node
2019-07-10T10:01:40.859291-04:00 sles15 dockerd[25407]: time="2019-07-10T10:01:40.859070236-04:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" module=node

Please help me in docker swarm join issue.
